fiat/coingecko.go

```diff
@@ -0,0 +1,276 @@
+// If timestamp is nil, it fetches the latest market data available.
+func (cg *CoingeckoDownloader) GetMarketData(timestamp *time.Time) ([]byte, error) {
+	requestURL := cg.url + "/coins/" + cg.coin
+	if timestamp != nil {
+		requestURL += "/history"
+	}
+
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		glog.Errorf("Error creating a new request for %v: %v", requestURL, err)
+		return nil, err
+	}
+	req.Close = true
+	req.Header.Set("Content-Type", "application/json")
+
+	// Add query parameters
+	q := req.URL.Query()
+	if timestamp == nil {
+		q.Add("market_data", "true")
+		q.Add("localization", "false")
+		q.Add("tickers", "false")
+		q.Add("community_data", "false")
+		q.Add("developer_data", "false")
+	} else {
+		timestampFormatted := timestamp.Format(cg.timeFormat)
+		q.Add("date", timestampFormatted)
+	}
+	req.URL.RawQuery = q.Encode()
+
+	client := &http.Client{
+		Timeout: cg.httpTimeoutSeconds,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New("Invalid response status: " + string(resp.Status))
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return bodyBytes, nil
+}
+
+// GetData gets fiat rates from coingecko.com at the specified date and returns JSON string.
+// If timestamp is nil, it will download the latest market data.
+func (cg *CoingeckoDownloader) GetData(timestamp *time.Time) (string, error) {
+	bodyBytes, err := cg.GetMarketData(timestamp)
+	if err != nil {
+		return "", err
+	}
+
+	type CoinGeckoResponse struct {
+		MarketData struct {
+			Prices map[string]interface{} `json:"current_price"`
+		} `json:"market_data"`
+	}
+
+	var data CoinGeckoResponse
+	err = json.Unmarshal(bodyBytes, &data)
+	if err != nil {
+		glog.Errorf("Error parsing CoinGecko response: %v", err)
+		return "", err
+	}
+
+	jsonString, err := json.Marshal(data.MarketData.Prices)
+	if err != nil {
+		glog.Errorf("Error marshalling CoinGecko prices: %v", err)
+		return "", err
+	}
+	return string(jsonString), nil
+}
+
+// MarketDataExists checks if there's data available for the specific timestamp.
+func (cg *CoingeckoDownloader) MarketDataExists(timestamp *time.Time) (bool, error) {
+	resp, err := cg.GetMarketData(timestamp)
+	if err != nil {
+		glog.Error("Error getting market data: ", err)
+		return false, err
+	}
+	type CoinGeckoResponse struct {
+		MarketData struct {
+			Prices map[string]interface{} `json:"current_price"`
+		} `json:"market_data"`
+	}
+	var data CoinGeckoResponse
+	err = json.Unmarshal(resp, &data)
+	if err != nil {
+		glog.Errorf("Error parsing CoinGecko response: %v", err)
+		return false, err
+	}
+	return len(data.MarketData.Prices) != 0, nil
+}
+
+// FindEarliestMarketData uses binary search to find the oldest market data available on CoinGecko.
+func (cg *CoingeckoDownloader) FindEarliestMarketData() (*time.Time, error) {
```

Comment:
Review in the your whole code what methods should be exported from package (uppercase) and all the other s make package only

Reply:
Fixed in https://github.com/trezor/blockbook/pull/316/commits/184380ec4eafdae606a402a9633b72e6e95a2dcf